### PR TITLE
Fix option unset checking.

### PR DESCRIPTION
**Pull request by halfflat**
* It turns out '0' is also a valid representation for unset in the `options` table. The code backing the options table doesn't check if the option is _really_ set, unlike the json cli options serializer. Some might assert that this does not constitute best design practice in a lua plugin interface.

### CODE REVIEW

**Comment by halfflat** [comment]
I thought I had tested all the previous patches together, but I must have missed a deploy, because with the change to the 'cpu' options check, salloc would barf. This should undo the damage.

**Review by pelahi — COMMENTED** [review]
From my knowledge I think something could also have a "?" or "0?" as a value. But that could no longer be the case as I think I encountered these values in 21.x. Do you know a way of testing all unset values? Maybe we should add something to get all the options and see what values they have.

File: luas/cli_filter.lua [hunk 1]
```diff
@@ -191,8 +191,10 @@ function slurm_cli_pre_submit(options, offset)
         return partition == 'gpu' or partition == 'gpu-dev' or partition == 'gpu-highmem'
     end
 
-    -- An unset option can be repesented by nil, the string "-2", or the string "unset": check all of them.
-    local function is_unset(x) return x == nil or x == '-2' or x == 'unset' end
+    -- An unset option can be repesented by nil, the string "-2", the string "0", or the string "unset": check all of them.
+    -- (Use explicit checks if any of the above is in fact an expected possible value. Note that options exposed by
+    -- the slurm cli_filter lua API differ between the json reporting and between the C-backed lua 'options' table.)
+    local function is_unset(x)  return x == nil or x == '-2' or x == 'unset' or x == '0' end
```
Comment (pelahi) [inline luas/cli_filter.lua]: Can we change this to an array (possibly global) that has the list of all values that x could take and be considered "unset"
```lua
local function in_table (tab, val)
  for index, value in ipairs(tab) do
    if value == val then
      return true
    end
 end
  return false
end
unset_values = {nil, '-2', 'unset', '0'}
local function is_unset(x) return in_table(unset_values, x) end
```

Comment (pelahi) [reply luas/cli_filter.lua]: Having it somewhere clearly defined might make it easier to update this list. Thoughts?

**Comment by halfflat** [comment]
> From my knowledge I think something could also have a "?" or "0?" as a value. But that could no longer be the case as I think I encountered these values in 21.x. Do you know a way of testing all unset values? Maybe we should add something to get all the options and see what values they have.

There is proper support for this inside slurm, and it's what is used to produce the json serialisation of the options set — each option has, potentially, its own specific dummy value representing an unset state, but also a pointer to a function that tests this state and reports set or unset. But the slurm-backed options table in the Lua interface doesn't bother checking against that interface, and so uncritically reports dummy values in the table.

Without modifying the slurm source, I think the only way to robustly test if an option is set or not is to scan the corresponding json string for keys. That said, I don't know strictly JSON the slurm serialisation code is, so there is still potentially some room for edge cases from misparsing.

**Comment by halfflat** [comment]
In the interim I'll rebase this patch on main so it can be merged from the PR.

**Review by pelahi — APPROVED** [review]
